File: VB/PersistentObjects.vb
Imports DevExpress.Xpo
Imports System

Namespace Northwind
    <Persistent("Customers")> _
    Public Class Customer
        Inherits XPLiteObject

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub
        Private _City As String
        Private _Country As String
        Private _ContactName As String
        Private _ContactTitle As String
        Private _CompanyName As String
        Private _CustomerId As String
        <Key> _
        Public Property CustomerId() As String
            Get
                Return _CustomerId
            End Get
            Set(ByVal value As String)
                SetPropertyValue("CustomerId", _CustomerId, value)
            End Set
        End Property
        Public Property CompanyName() As String
            Get
                Return _CompanyName
            End Get
            Set(ByVal value As String)
                SetPropertyValue("CompanyName", _CompanyName, value)
            End Set
        End Property
        Public Property ContactTitle() As String
            Get
                Return _ContactTitle
            End Get
            Set(ByVal value As String)
                SetPropertyValue("ContactTitle", _ContactTitle, value)
            End Set
        End Property
        Public Property ContactName() As String
            Get
                Return _ContactName
            End Get
            Set(ByVal value As String)
                SetPropertyValue("ContactName", _ContactName, value)
            End Set
        End Property
        Public Property Country() As String
            Get
                Return _Country
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Country", _Country, value)
            End Set
        End Property
        Public Property City() As String
            Get
                Return _City
            End Get
            Set(ByVal value As String)
                SetPropertyValue("City", _City, value)
            End Set
        End Property
        <Association("CustomerOrders", GetType(Order))> _
        Public ReadOnly Property Orders() As XPCollection(Of Order)
            Get
                Return GetCollection(Of Order)("Orders")
            End Get
        End Property
    End Class

    <Persistent("Orders")> _
    Public Class Order
        Inherits XPLiteObject

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Private _Freight As Decimal
        Private _Employee As Employee
        Private _Customer As Customer
        Private _ShippedDate As Date
        Private _OrderID As Integer
        <Key> _
        Public Property OrderID() As Integer
            Get
                Return _OrderID
            End Get
            Set(ByVal value As Integer)
                SetPropertyValue("OrderID", _OrderID, value)
            End Set
        End Property
        Public Property ShippedDate() As Date
            Get
                Return _ShippedDate
            End Get
            Set(ByVal value As Date)
                SetPropertyValue("ShippedDate", _ShippedDate, value)
            End Set
        End Property
        <Persistent("CustomerID"), Association("CustomerOrders")> _
        Public Property Customer() As Customer
            Get
                Return _Customer
            End Get
            Set(ByVal value As Customer)
                SetPropertyValue("Customer", _Customer, value)
            End Set
        End Property

        <Persistent("EmployeeID"), Association("EmployeeOrders")> _
        Public Property Employee() As Employee
            Get
                Return _Employee
            End Get
            Set(ByVal value As Employee)
                SetPropertyValue("Employee", _Employee, value)
            End Set
        End Property
        Public Property Freight() As Decimal
            Get
                Return _Freight
            End Get
            Set(ByVal value As Decimal)
                SetPropertyValue("Freight", _Freight, value)
            End Set
        End Property
    End Class
    <Persistent("Employees")> _
    Public Class Employee
        Inherits XPLiteObject

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub

        Private _LastName As String
        Private _FirstName As String
        Private _EmployeeID As Integer
        <Key> _
        Public Property EmployeeID() As Integer
            Get
                Return _EmployeeID
            End Get
            Set(ByVal value As Integer)
                SetPropertyValue("EmployeeID", _EmployeeID, value)
            End Set
        End Property
        Public Property FirstName() As String
            Get
                Return _FirstName
            End Get
            Set(ByVal value As String)
                SetPropertyValue("FirstName", _FirstName, value)
            End Set
        End Property
        Public Property LastName() As String
            Get
                Return _LastName
            End Get
            Set(ByVal value As String)
                SetPropertyValue("LastName", _LastName, value)
            End Set
        End Property
        <Association("EmployeeOrders")> _
        Public ReadOnly Property Orders() As XPCollection(Of Order)
            Get
                Return GetCollection(Of Order)("Orders")
            End Get
        End Property
    End Class
End Namespace

File: CS/E642/PersistentObjects.cs
namespace Northwind {
    using System;
    using DevExpress.Xpo;
    [Persistent("Customers")]
    public class Customer : XPLiteObject {
        public Customer(Session session) : base(session) { }
        private string _City;
        private string _Country;
        private string _ContactName;
        private string _ContactTitle;
        private string _CompanyName;
        private string _CustomerId;
        [Key]
        public string CustomerId {
            get {
                return _CustomerId;
            }
            set {
                SetPropertyValue("CustomerId", ref _CustomerId, value);
            }
        }
        public string CompanyName {
            get {
                return _CompanyName;
            }
            set {
                SetPropertyValue("CompanyName", ref _CompanyName, value);
            }
        }
        public string ContactTitle {
            get {
                return _ContactTitle;
            }
            set {
                SetPropertyValue("ContactTitle", ref _ContactTitle, value);
            }
        }
        public string ContactName {
            get {
                return _ContactName;
            }
            set {
                SetPropertyValue("ContactName", ref _ContactName, value);
            }
        }
        public string Country {
            get {
                return _Country;
            }
            set {
                SetPropertyValue("Country", ref _Country, value);
            }
        }
        public string City {
            get {
                return _City;
            }
            set {
                SetPropertyValue("City", ref _City, value);
            }
        }
        [Association("CustomerOrders", typeof(Order))]
        public XPCollection<Order> Orders {
            get {
                return GetCollection<Order>("Orders");
            }
        }
    }

    [Persistent("Orders")]
    public class Order : XPLiteObject {
        public Order(Session session) : base(session) { }

        private decimal _Freight;
        private Employee _Employee;
        private Customer _Customer;
        private DateTime _ShippedDate;
        private int _OrderID;
        [Key]
        public int OrderID {
            get {
                return _OrderID;
            }
            set {
                SetPropertyValue("OrderID", ref _OrderID, value);
            }
        }
        public DateTime ShippedDate {
            get {
                return _ShippedDate;
            }
            set {
                SetPropertyValue("ShippedDate", ref _ShippedDate, value);
            }
        }
        [Persistent("CustomerID"), Association("CustomerOrders")]
        public Customer Customer {
            get {
                return _Customer;
            }
            set {
                SetPropertyValue("Customer", ref _Customer, value);
            }
        }

        [Persistent("EmployeeID"), Association("EmployeeOrders")]
        public Employee Employee {
            get {
                return _Employee;
            }
            set {
                SetPropertyValue("Employee", ref _Employee, value);
            }
        }
        public decimal Freight {
            get {
                return _Freight;
            }
            set {
                SetPropertyValue("Freight", ref _Freight, value);
            }
        }
    }
    [Persistent("Employees")]
    public class Employee : XPLiteObject {
        public Employee(Session session) : base(session) { }

        private string _LastName;
        private string _FirstName;
        private int _EmployeeID;
        [Key]
        public int EmployeeID {
            get {
                return _EmployeeID;
            }
            set {
                SetPropertyValue("EmployeeID", ref _EmployeeID, value);
            }
        }
        public string FirstName {
            get {
                return _FirstName;
            }
            set {
                SetPropertyValue("FirstName", ref _FirstName, value);
            }
        }
        public string LastName {
            get {
                return _LastName;
            }
            set {
                SetPropertyValue("LastName", ref _LastName, value);
            }
        }
        [Association("EmployeeOrders")]
        public XPCollection<Order> Orders {
            get {
                return GetCollection<Order>("Orders");
            }
        }
    }
}